File: 1696-jump-game-vi/1696-jump-game-vi.go
func maxResult(nums []int, k int) int {
    var deq []int
    deq = append(deq, len(nums) - 1)
    
    for i := len(nums) - 2; i >= 0; i-- {
        if deq[0] - i > k {
            deq = deq[1:]
        }
        
        nums[i] += nums[deq[0]]
        
        for len(deq) > 0 && nums[deq[len(deq) - 1]] <= nums[i] {
            deq = deq[:len(deq) - 1]
        }
        
        deq = append(deq, i)
    }
    
    return nums[0]
}

File: 1696-jump-game-vi/1696-jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        deq = deque()
        deq.append(len(nums) - 1)
        
        for i in range(len(nums) - 1)[::-1]:
            if (deq[0] - i > k):
                deq.popleft()
            
            nums[i] += nums[deq[0]]
            
            while (deq and nums[deq[-1]] <= nums[i]):
                deq.pop()
            
            deq.append(i)
        
        return nums[0]